### DISCUSSION TITLE
Incremental testing - seconds not hours - AppMaps + Solidus

### DISCUSSION THREAD

**Discussion by kgilpin** [solidusio/solidus#4000] [Show and tell]
Hi everyone, we've been continuing to advance our open source project [AppMap](https://marketplace.visualstudio.com/items?itemName=appland.appmap), frequently using Solidus as an example project in our work. 

One of the new capabilities that we are developing is incremental testing. As you code, you can run a rake task like `rake rspec:incremental`, and this rake task examines your source code changes and runs just the subset of test cases that are likely affected by your code changes. It's like using [Guard](https://github.com/guard/guard-rspec#typical-rails-app),
except that unlike Guard, it's not based on file patterns, it's based on actual usage of code within test case execution. So, the dependency detection is a lot more powerful. 

Here's a video of the workflow using Solidus. 

<a href="https://www.loom.com/share/d6008ed6916e45d8a07443bef309347f"> <p>Incremental diff demo - March 16 2021 - Watch Video</p> <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/d6008ed6916e45d8a07443bef309347f-with-play.gif"> </a>

(If you are super impatient, or you have seen AppMap before, skip ahead to `2:21`. Otherwise, you may find the whole video interesting).

**If you're interested in trying this out** in your development work (with Solidus, or in another Ruby, Java, or Python project), please let me know since we are actively looking for early users and feedback on this open source tool! Thanks for your time.

**Comment by kennyadsl** [comment]
Interesting, thanks for sharing! Does it also work between subcomponents of Solidus? Eg. some change in the `core` component also triggers related specs from `frontend` or `backend` that depend on the code changed?

**Reply by kgilpin** [reply]
Hi, yes it can work that way. Here's an example I just recorded in which I change a file in "core" and run dependent tests in "backend". I think this could be smoothed out and improved quite a bit more, but you can get the basic idea :-)

https://www.loom.com/share/824c6ca2a55e4c969c90fbf50ee91533

**Reply by kennyadsl** [reply]
And specs pass despite that change, they shouldn't! Of course, it's due to how specs are written, not to this tool! 🤦
